sys-boot/{refind,udk}: Bugs 588440, 566930, 563226 and 582716

sveyret opened the pull request:
Here is the PR containing the patchs uploaded to Bugzilla for the indicated bugs.
@wraeth

Amynka commented:
@gentoo/proxy-maint

SoapGentoo commented on sys-boot/refind/refind-0.10.2.ebuild:
```diff
@@ -39,19 +40,78 @@ pkg_setup() {
 }
 
+src_prepare() {
+	eapply_user
```
prefer `default` here

SoapGentoo commented on sys-boot/refind/refind-0.10.2.ebuild:
```diff
@@ -39,19 +40,78 @@ pkg_setup() {
 
+src_prepare() {
+	eapply_user
+	for f in "${S}"/*/Make.tiano "${S}"/Make.common; do
```
declare `f` local before to keep environmental pollution to a minimum

SoapGentoo commented on sys-boot/refind/refind-0.10.2.ebuild:
```diff
@@ -39,19 +40,78 @@ pkg_setup() {
 	# Make filesystem drivers
-	export gnuefi_target="_gnuefi"
+	use gnuefi && export gnuefi_target="_gnuefi"
 	for fs in ${FS_USE}; do
```
again, declare `local` before

SoapGentoo commented on sys-boot/udk/udk-2015.ebuild:
```diff
@@ -0,0 +1,241 @@
+	if use doc; then
+		mkdir -p "${S}/doc" || die
+		pushd "${S}/doc" >/dev/null || die
+		for f in "${WORKDIR}/Documents/"*" Document.zip"; do
```
make `local`

gktrk commented on sys-boot/refind/refind-0.10.2.ebuild:
```diff
@@ -39,19 +40,78 @@ pkg_setup() {
+			"${f}" || die "Failed to patch Tianocore make file in" \
+			$(basename $(dirname ${f}))
+	done
+	sed -i -e '/Guids/i#include "AutoGen.h"\n' "${S}/filesystems/AutoGen.c"
```
missing ||die here

gktrk commented on sys-boot/refind/refind-0.10.2.ebuild:
```diff
@@ -39,19 +40,78 @@ pkg_setup() {
+	done
+	sed -i -e '/Guids/i#include "AutoGen.h"\n' "${S}/filesystems/AutoGen.c"
+	for f in "${S}"/*/AutoGen.c; do
+		cat >>"${f}" <<EOF
```
missing ||die here

sveyret commented:
Thank you all for your review. Just a short comment to warn you that I have an Internet breakdown for 2 weeks now and don't know when I will be able to work on it again.

sveyret commented:
Hi,
Your suggestions are now included in the pull request.
Thank you!

SoapGentoo commented on sys-boot/refind/refind-0.10.2.ebuild:
```diff
@@ -39,19 +40,84 @@ pkg_setup() {
+	sed -i -e '/Guids/i#include "AutoGen.h"\n' "${S}/filesystems/AutoGen.c" \
+		|| die "Failed to patch AutoGen.c"
+	for f in "${S}"/*/AutoGen.c; do
+		(cat >>"${f}" <<EOF
```
please just add the `|| die` after the EOF instead.

SoapGentoo commented on sys-boot/udk/udk-2015.ebuild:
```diff
@@ -0,0 +1,242 @@
+}
+
+src_configure() {
+	python_setup 'python2.7'
```
shouldn't this happen in `pkg_setup()` ?

SoapGentoo commented on sys-boot/udk/udk-2015.ebuild:
```diff
@@ -0,0 +1,242 @@
+src_install() {
+	local build_dir="${S}/Build/MdeModule/RELEASE_${TOOLCHAIN_TAG}/${ARCH}"
+
+	for f in "${build_dir}"/*/Library/*/*/OUTPUT/*.lib; do
```
make f `local`

sveyret commented:
@SoapGentoo Thank you, all done…

gktrk commented:
The build is failing for me:

```
x86_64-pc-linux-gnu-ld -o refind_x64.dll -n -q --gc-sections -nostdlib 			--script=/usr/lib/GccBase.lds 			--defsym=PECOFF_HEADER_SIZE=0x228 			--entry efi_main -u efi_main -m elf_x86_64 \
	  -L /usr/lib --start-group -lBaseDebugPrintErrorLevelLib -lBasePrintLib -lBasePcdLibNull -lUefiDebugLibStdErr -lBaseLib -lBaseMemoryLib -lUefiBootServicesTableLib -lUefiMemoryAllocationLib -lUefiRuntimeServicesTableLib -lUefiDevicePathLib -lUefiLib -lPeCoffExtraActionLibNull -lUefiApplicationEntryPoint -lDxeServicesLib -lDxeServicesTableLib -lDxeHobLib -lBasePeCoffLib apple.obj config.obj mystrings.obj line_edit.obj driver_support.obj icns.obj lib.obj main.obj menu.obj screen.obj gpt.obj crc32.obj legacy.obj AutoGen.obj ../libeg/libeg.lib \
	  ../EfiLib/EfiLib.lib ../mok/mok.lib --end-group
../libeg/libeg.lib(lodepng.obj): In function `lodepng_state_copy':
lodepng.c:(.text+0x4f65): undefined reference to `memcpy'
Make.tiano:48: recipe for target 'refind_x64.dll' failed
make[1]: *** [refind_x64.dll] Error 1
make[1]: Leaving directory '/var/tmp/portage/sys-boot/refind-0.10.3/work/refind-0.10.3/refind'
Makefile:54: recipe for target 'tiano' failed
make: *** [tiano] Error 2
```

First I thought removing `-nostdlib` may help, but it didn't. Am I missing something?

sveyret commented:
@gktrk I will investigate, but I don't have this error, although I've got the exact same command executed. Which USE did you set ? Default ones ?

sveyret commented:
@gktrk So, there is indeed a call to memcpy in lodepng.c, but it should not stay after compilation as memcpy is a #define in lodepng.h
Do you have a compilation option which would prevent from taking the good header file, or which would ignore the #define?

sveyret commented:
So, after discussions in gentoo-proxy-maint channel, we have to try to replace #include string.h with stddef.h.

gktrk commented:
shall we sed the sources on our end and file a bug upstream?

gktrk commented:
let me know if you need my build logs

sveyret commented:
@gktrk That's what I'll try. But as I don't have the problem in my machin, that would be good if you could do a test by your side so I'd be sure it works like that…

gktrk commented:
sure, I'll get the logs, and also patch it and runtest tonight

gktrk commented:
Interestingly, it did not fix the issue. I appended the following to src_prepare():
```
sed -i -e 's|<string.h>|<stddef.h>|' libeg/lodepng.h || die
```

but the build still fails. If I cd into ${WORKDIR}, do make clean && make it compiles file. Here's how we compile lodepng.o:

```
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -fno-strict-aliasing -fno-stack-protector -fshort-wchar -Wall -DEFIX64 -DEFI_FUNCTION_WRAPPER -m64 -mno-red-zone  -fpic -I/usr/include/efi -I/usr/include/efi/x86_64 -I/usr/include/efi/protocol -I../include -I../refind -I../libeg -I../mok -I. -I./../include \
      -D__MAKEWITH_GNUEFI -DREFIND_VERSION='L"0.10.3"' -c lodepng.c -o lodepng.o
```

Here's the line after make clean && make:

```
/usr/bin/gcc -Os -fno-strict-aliasing -fno-stack-protector -fshort-wchar -Wall -DEFIX64 -DEFI_FUNCTION_WRAPPER -m64 -mno-red-zone  -fpic -I/usr/include/efi -I/usr/include/efi/x86_64 -I/usr/include/efi/protocol -I../include -I../refind -I../libeg -I../mok -I. -I./../include \
      -D__MAKEWITH_GNUEFI -DREFIND_VERSION='L"0.10.3"' -c lodepng.c -o lodepng.o
```

Here's our linker line:

```
x86_64-pc-linux-gnu-ld -L./../libeg/ -L./../mok/ -L./../EfiLib/ -T /usr/lib/elf_x86_64_efi.lds -shared -nostdlib -Bsymbolic 		-L/usr/lib -L/usr/lib /usr/lib/crt0-efi-x86_64.o -znocombreloc -zdefs  main.o mystrings.o apple.o line_edit.o config.o menu.o screen.o icns.o gpt.o crc32.o lib.o driver_support.o legacy.o \
      -o refind_x64.so -leg -lmok -lEfiLib -lefi -lgnuefi /usr/lib/gcc/x86_64-pc-linux-gnu/4.9.3/libgcc.a
```

Here's the linker line with make clean && make:

```
/usr/bin/ld -L./../libeg/ -L./../mok/ -L./../EfiLib/ -T /usr/lib/elf_x86_64_efi.lds -shared -Bsymbolic -nostdlib -L/usr/lib -L/usr/lib /usr/lib/crt0-efi-x86_64.o -znocombreloc -zdefs  main.o mystrings.o apple.o line_edit.o config.o menu.o screen.o icns.o gpt.o crc32.o lib.o driver_support.o legacy.o \
      -o refind_x64.so -leg -lmok -lEfiLib -lefi -lgnuefi /usr/lib/gcc/x86_64-pc-linux-gnu/4.9.3/libgcc.a
```

I have the following in `make.conf`:

```
CFLAGS="-O2 -pipe -march=native"
```

If I replace it with:

```
CFLAGS="-Os -pipe -march=native"
```

the compilation succeeds. Perhaps `-Os` is explicitly required?

sveyret commented:
If I am not mistaking, my compilation line is exactly same as yours:
`x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -fno-strict-aliasing -fno-stack-protector -fshort-wchar -Wall -DEFIX64 -DEFI_FUNCTION_WRAPPER -m64 -mno-red-zone  -mcmodel=large "-DEFIAPI=__attribute__((ms_abi))"  -I /usr/include/udk -I .. -I ../refind -I ../libeg -I ../include -I ../mok -DNO_BUILTIN_VA_FUNCS \
      -D__MAKEWITH_TIANO -DREFIND_VERSION='L"0.10.3"' -c lodepng_xtra.c -o lodepng_xtra.obj
`
…and it's working! I can change the ebuild to force size optimization, if it really is an issue. Do you think I shall try that?

gktrk commented:
We shouldn't needlessly force flags. I'll debug it some more tonight and test it on a different machine. Perhaps the problem is isolated to my setup.

gktrk commented:
I can reproduce the error on a different machine which has `gcc version 4.9.3 (Gentoo 4.9.3 p1.4, pie-0.6.4)`. If I switch to `gcc version 4.8.5 (Gentoo 4.8.5 p1.3, pie-0.6.2)` on the same machine, it compiles fine.

gktrk commented:
Issue persists with `gcc version 4.9.4 (Gentoo 4.9.4 p1.0, pie-0.6.4)`

doughdemon commented:
The upstream commit https://sourceforge.net/p/refind/code/ci/430a0cb5b65d028351f2aa71f3bf3d225803d1fb/ looks relevant.

sveyret commented:
For information, my version of gcc is `gcc (Gentoo 4.9.3 p1.5, pie-0.6.4) 4.9.3`. I'm not sure I'll have time to investigate the coming days, but I'd like to understand what the problem is… :-/

sveyret commented:
After researching information on the web, the only thing I can think of is a problem of optimisation done by gcc, which seems to be confirmed by the options you tried to make it work. If I have time this week-end, I will try to disactivate optimisation for this file compilation.

gktrk commented:
Compiles fine with `-O2` using `gcc version 4.10.0-pre9999 20160920 (experimental) commit ae96b054541acc606d6a9401aa4551f5ef53a21d (Gentoo 4.10.0_pre9999)`. Maybe we hit a regression?

gktrk commented:
Somehow -O2 inserts a memcpy there during object creation:

```
        movq    %rbp, %rsi
-       rep movsl
-       leaq    80(%rbx), %r12
-       movq    %r13, %rdi
-       movabsq $lodepng_color_mode_init, %rax
-       call    *%rax
-       movq    %r12, %rdi
-       movabsq $lodepng_info_init, %rax
+       movq    %rbx, %rdi
+       movabsq $memcpy, %rax
        call    *%rax
```

Looks like a known gcc issue: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56888

gktrk commented:
Reverting the commit in https://sourceforge.net/p/refind/code/ci/430a0cb5b65d028351f2aa71f3bf3d225803d1fb/ fixes it for me

sveyret commented:
The commit was done for a compilation problem in ARM, so I think it is not a problem to revert for amd64/x86.
Solution 1: sed to simulate code revert if amd64 or x86.
Solution 2: sed to add a #if which ignore the #define if processor is amd64 or x86.
Solution 3: force -Os if gcc version is 4.9.x.
What do you think is best?

gktrk commented:
Wish we know what the exact problem with arm was. I checked the preprocessor output of the file. Each reference to `memcpy` is successfully replaced with `MyMemCpy`. If I understand correctly, gcc emits extra `memcpy` during the optimization stage if it deems helpful. These are inserted past the preprocessing, so they are not converted properly with the `#define`. Once it hits the linking stage, because of `-nostdlib`, it fails. This wasn't a problem before because refind redefined `memcpy` instead of the redirection to `MyMemCpy` and the `memcpy` inserted by gcc are resolved correctly.

Refind isn't keyworded for `arm` at the moment, so we can just revert the code unconditionally. I should probably file an upstream bug about the issue.

gktrk commented:
And we should revert with a patch instead of using sed to keep it more coherent. Bunch of sed statements will make little sense in the ebuild.

sveyret commented:
@gktrk I made the patch, if you want to process a patch. I will also file a bug upstream.

gktrk commented:
Sorry I am on vacation but I will test this in a day or two.

gktrk commented:
Compiles fine with the our last patch

gktrk commented:
So, I'm testing this on a hardened profile and udk fails for me with the following:

```
x86_64-pc-linux-gnu-gcc -o ../bin/GnuGenBootSector  GnuGenBootSector.o -L../libs -lCommon
/usr/lib/gcc/x86_64-pc-linux-gnu/4.9.3/../../../../x86_64-pc-linux-gnu/bin/ld: GnuGenBootSector.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
GnuGenBootSector.o: error adding symbols: Bad value
collect2: error: ld returned 1 exit status
../Makefiles/app.makefile:24: recipe for target '../bin/GnuGenBootSector' failed
make[2]: *** [../bin/GnuGenBootSector] Error 1
```

If I patch like the following, it compiles fine. Didn't get a chance to take a deeper look yet.

```
diff --git a/sys-boot/udk/udk-2015.ebuild b/sys-boot/udk/udk-2015.ebuild
index 5452d5f..cdbb6ad 100644
--- a/sys-boot/udk/udk-2015.ebuild
+++ b/sys-boot/udk/udk-2015.ebuild
@@ -72,6 +72,7 @@ src_configure() {
        append-cflags $(test-flags-CC -MD) $(test-flags-CC -fshort-wchar)
        append-cflags $(test-flags-CC -fno-strict-aliasing)
        append-cflags $(test-flags-CC -nostdlib) $(test-flags-CC -c)
+       append-cflags -fPIC
        sed -e "s:^\(CFLAGS\s*=\).*$:\1 ${CFLAGS}:" \
                -i "${S}/BaseTools/Source/C/Makefiles/header.makefile" \
                || die "Failed to update makefile header"
```

And refind fails with the following:

```
x86_64-pc-linux-gnu-ld -o refind_x64.dll -n -q --gc-sections -nostdlib 			--script=/usr/lib/GccBase.lds 			--defsym=PECOFF_HEADER_SIZE=0x228 			--entry efi_main -u efi_main -m elf_x86_64 \
	  -L /usr/lib --start-group -lBaseDebugPrintErrorLevelLib -lBasePrintLib -lBasePcdLibNull -lUefiDebugLibStdErr -lBaseLib -lBaseMemoryLib -lUefiBootServicesTableLib -lUefiMemoryAllocationLib -lUefiRuntimeServicesTableLib -lUefiDevicePathLib -lUefiLib -lPeCoffExtraActionLibNull -lUefiApplicationEntryPoint -lDxeServicesLib -lDxeServicesTableLib -lDxeHobLib -lBasePeCoffLib apple.obj config.obj mystrings.obj line_edit.obj driver_support.obj icns.obj lib.obj main.obj menu.obj screen.obj gpt.obj crc32.obj legacy.obj AutoGen.obj ../libeg/libeg.lib \
	  ../EfiLib/EfiLib.lib ../mok/mok.lib --end-group
line_edit.obj: In function `line_edit':
line_edit.c:(.text+0x475): undefined reference to `memmove'
line_edit.c:(.text+0x530): undefined reference to `memmove'
line_edit.c:(.text+0x6e6): undefined reference to `memmove'
Make.tiano:48: recipe for target 'refind_x64.dll' failed
make[1]: *** [refind_x64.dll] Error 1
make[1]: Leaving directory '/var/tmp/portage/sys-boot/refind-0.10.3/work/refind-0.10.3/refind'
Makefile:54: recipe for target 'tiano' failed
make: *** [tiano] Error 2
```

Working on that now.

sveyret commented:
After reading some docs on gcc optimization, I added the -fno-tree-loop-distribute-patterns flag to gcc which prevents gcc from replacing some loops with standard memcpy, memmove, memset functions.
I think I managed to reproduice the problem and this flag corrects it. I modified my commits in order to use this new method. Please tell me if everything is working now.
P.S. Will work on 0.10.4 after that.

mgorny commented:
@sveyret, the gcc bug report actually talks of `-ffreestanding` flag. AFAIU it prevents the compiler from generating any builtin function calls. Isn't that what you want here?

sveyret commented:
@mgorny If I understood well, even a freestanding system must have a memcpy, memmove, etc. functions, and so gcc is replacing code by calls to these functions during its optimization process. In the issue given by @gktrk , the reporter says he *is* using -ffreestanding and has the problem anyway.

gktrk commented:
The sed in src_unpack of udk fails with:

```
Appending /home/gokturk/stuff/gentoo to PORTDIR_OVERLAY...
 * UDK2015.Complete.MyWorkSpace.zip SHA256 SHA512 WHIRLPOOL size ;-) ... [ ok ]
>>> Unpacking source...
>>> Unpacking UDK2015.Complete.MyWorkSpace.zip to /var/tmp/portage/sys-boot/udk-2015/work
>>> Unpacking /var/tmp/portage/sys-boot/udk-2015/work/UDK2015.MyWorkSpace.zip to /var/tmp/portage/sys-boot/udk-2015/work
/var/tmp/portage/sys-boot/udk-2015/work/MyWorkSpace /var/tmp/portage/sys-boot/udk-2015/work
>>> Unpacking /var/tmp/portage/sys-boot/udk-2015/work/BaseTools(Unix).tar to /var/tmp/portage/sys-boot/udk-2015/work/MyWorkSpace
sed: -e expression #1, char 31: invalid reference \1 on `s' command's RHS
sed: -e expression #1, char 31: invalid reference \1 on `s' command's RHS
sed: -e expression #1, char 31: invalid reference \1 on `s' command's RHS
sed: -e expression #1, char 31: invalid reference \1 on `s' command's RHS
sed: -e expression #1, char 31: invalid reference \1 on `s' command's RHS
>>> Source unpacked in /var/tmp/portage/sys-boot/udk-2015/work
```

It's missing the escapes on `( )` for the submatch:

```
 sys-boot/udk/udk-2015.ebuild | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sys-boot/udk/udk-2015.ebuild b/sys-boot/udk/udk-2015.ebuild
index b758cf7..b3023dd 100644
--- a/sys-boot/udk/udk-2015.ebuild
+++ b/sys-boot/udk/udk-2015.ebuild
@@ -53,7 +53,7 @@ src_unpack() {
 		mkdir -p "${S}/doc" || die
 		pushd "${S}/doc" >/dev/null || die
 		for f in "${WORKDIR}/Documents/"*" Document.zip"; do
-			doc_name=$(echo ${f} | sed -e 's:^.*/([^/]*) Document.zip$:\1:')
+			doc_name=$(echo ${f} | sed -e 's:^.*/\([^/]*\) Document.zip$:\1:')
 			if [[ -f "${WORKDIR}/Documents/${doc_name} Document.zip" ]]; then
 				unpack "${WORKDIR}/Documents/${doc_name} Document.zip"
 				mv "${S}/doc/html" "${S}/doc/${doc_name}" || die
```

will fix it here before commit

gktrk commented:
should also put that `.` before `zip` in `[ ]` for an exact match on `.`

gktrk commented:
(drum rolls ...)

Merged in https://github.com/gentoo/gentoo/compare/ffcd38e...431a8c3! Thanks for being patient with us and apologies for the crazy delay.
